DependencyInjection/Compiler/CacheCompatibilityPass.php

```diff
@@ -29,44 +31,84 @@ public function process(ContainerBuilder $container): void
     {
         foreach (array_keys($container->findTaggedServiceIds(self::CONFIGURATION_TAG)) as $id) {
             foreach ($container->getDefinition($id)->getMethodCalls() as $methodCall) {
+                if ($methodCall[0] === 'setSecondLevelCacheConfiguration') {
+                    $slcConfigDefinition = $methodCall[1][0];
+                    $this->updateSecondLevelCache($container, $slcConfigDefinition);
+                    continue;
+                }
+
                 if (! isset(self::CACHE_METHODS_PSR6_SUPPORT_MAP[$methodCall[0]])) {
                     continue;
                 }
 
                 $aliasId      = (string) $methodCall[1][0];
                 $definitionId = (string) $container->getAlias($aliasId);
-                $definition   = $container->getDefinition($definitionId);
                 $shouldBePsr6 = self::CACHE_METHODS_PSR6_SUPPORT_MAP[$methodCall[0]];
 
-                while (! $definition->getClass() && $definition instanceof ChildDefinition) {
-                    $definition = $container->findDefinition($definition->getParent());
-                }
+                $this->wrapIfNecessary($container, $aliasId, $definitionId, $shouldBePsr6);
+            }
+        }
+    }
 
-                if ($shouldBePsr6 === is_a($definition->getClass(), CacheItemPoolInterface::class, true)) {
-                    continue;
-                }
+    private function updateSecondLevelCache(ContainerBuilder $container, Definition $slcConfigDefinition): void
+    {
+        foreach ($slcConfigDefinition->getMethodCalls() as $methodCall) {
+            if ($methodCall[0] !== 'setCacheFactory') {
+                continue;
+            }
 
-                $targetClass   = CacheProvider::class;
-                $targetFactory = DoctrineProvider::class;
+            $factoryDefinition = $methodCall[1][0];
+            assert($factoryDefinition instanceof Definition);
+            break;
+        }
 
-                if ($shouldBePsr6) {
-                    $targetClass   = CacheItemPoolInterface::class;
-                    $targetFactory = CacheAdapter::class;
+        $aliasId      = (string) $factoryDefinition->getArgument(1);
```

Comment:
From SA perspective, `$factoryDefinition` may not exist. Everything below this line, including this line, should be moved inside `foreach`, where existence of `$factoryDefinition` is guaranteed. I would suggest
```php
foreach ($slcConfigDefinition->getMethodCalls() as $methodCall) {      
    if ($methodCall[0] !== 'setCacheFactory') {
        continue;
    }
    $factoryDefinition = $methodCall[1][0];
    assert($factoryDefinition instanceof Definition);
    $aliasId = (string) $factoryDefinition->getArgument(1);
    $this->wrapIfNecessary($container, $aliasId, (string) $container->getAlias($aliasId), false);
    break;
}
```